File: web/web/database/versifications.php
<?php
/**
* @package bibledit
*/
/*
 ** Copyright (©) 2003-2012 Teus Benschop.
 **
 ** This program is free software; you can redistribute it and/or modify
 ** it under the terms of the GNU General Public License as published by
 ** the Free Software Foundation; either version 3 of the License, or
 ** (at your option) any later version.
 **  
 ** This program is distributed in the hope that it will be useful,
 ** but WITHOUT ANY WARRANTY; without even the implied warranty of
 ** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 ** GNU General Public License for more details.
 **  
 ** You should have received a copy of the GNU General Public License
 ** along with this program; if not, write to the Free Software
 ** Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 **  
 */


/**
* This is a database for the various versification systems.
*/
class Database_Versifications
{
  private static $instance;
  private function __construct() {
  } 
  public static function getInstance() 
  {
    if ( empty( self::$instance ) ) {
      self::$instance = new Database_Versifications();
    }
    return self::$instance;
  }


  public function optimize () {
    $database_instance = Database_Instance::getInstance();
    $database_instance->runQuery ("OPTIMIZE TABLE versification_names;");
    $database_instance->runQuery ("OPTIMIZE TABLE versification_data;");
  }


  public function importBibleditXml ($contents, $name)
  {
    // Delete old system if it is there, and create new one.
    $this->delete ($name);
    $id = $this->create ($name);
    $database_instance = Database_Instance::getInstance();
    $database_books = Database_Books::getInstance();
    $xml = simplexml_load_string ($contents);
    foreach ($xml as $triad) {
      $book = $triad->book;
      $book = $database_books->getIdFromEnglish ($book);
      $chapter = $triad->chapter;
      $verse = $triad->verse;
      $query = "INSERT INTO versification_data VALUES (NULL, $id, $book, $chapter, $verse);";
      $database_instance->runQuery ($query);
    }
  }


  public function exportBibleditXmlFile ($name)
  {
    $database_books = Database_Books::getInstance();
    $xml = new SimpleXMLElement ("<bibledit-versification-system></bibledit-versification-system>");
    $versification_data = $this->getBooksChaptersVerses ($name);
    while ($row = $versification_data->fetch_assoc()) {
      $triad = $xml->addChild ("triad");
      $book = $row["book"];
      $book = $database_books->getEnglishFromId ($book);     
      $triad->addChild ("book", $book);
      $chapter = $row["chapter"];
      $triad->addChild ("chapter", $chapter);
      $verse = $row["verse"];
      $triad->addChild ("verse", $verse);
    }
    $domnode = dom_import_simplexml ($xml);
    $dom = new DOMDocument ("1.0", "UTF-8");
    $domnode = $dom->importNode ($domnode, true);
    $dom->appendChild ($domnode);
    $dom->formatOutput = true;
    $string = $dom->saveXML ();
    $string = htmlentities ($string);
    return $string;
  }

  
  /**
    * delete - Delete a versification system.
    */      
  public function delete ($name)
  {
    $database_instance = Database_Instance::getInstance();
    $id = $this->getID ($name);
    $query = "DELETE FROM versification_names WHERE system = $id;";
    $database_instance->runQuery ($query);
    $query = "DELETE FROM versification_data WHERE system = $id;";
    $database_instance->runQuery ($query);
  }

  
  /**
    * getID - Returns the ID for a named versification system.
    */      
  public function getID ($name)
  {
    $database_instance = Database_Instance::getInstance();
    $name = Database_SQLInjection::no ($name);
    $query = "SELECT system FROM versification_names WHERE name = '$name';";
    $result = $database_instance->runQuery ($query);
    if ($result->num_rows == 0) {
      return 0;
    }
    $row = $result->fetch_row();
    return $row[0];
  }


  /**
    * create - Creates a new empty versification system.
    * Returns the new ID.
    */      
  public function create ($name)
  {
    // If the versification system already exists, return its ID.
    $id = $this->getID ($name);
    if ($id > 0) {
      return $id;
    }   
    // Get the first free ID starting from 1000.
    $database_instance = Database_Instance::getInstance();
    $query = "SELECT system FROM versification_names ORDER BY system DESC LIMIT 1;";
    $result = $database_instance->runQuery ($query);
    $id = 0;
    if ($result->num_rows > 0) {
      $row = $result->fetch_row();
      $id = $row[0];
    }
    $id++;
    if ($id < 1000) $id = 1000;
    // Create the empty system.
    $name = Database_SQLInjection::no ($name);
    $query = "INSERT INTO versification_names VALUES ($id, '$name');";
    $database_instance->runQuery ($query);
    // Return new ID.
    return $id;
  }


  /**
    * getSystems - Returns an array of the available versification systems.
    */      
  public function getSystems ()
  {
    $database_instance = Database_Instance::getInstance();
    $query = "SELECT name FROM versification_names ORDER BY name ASC;";
    $result = $database_instance->runQuery ($query);
    for ($i = 0; $i < $result->num_rows; $i++) {
      $row = $result->fetch_row();
      $systems[] = $row[0];
    }
    return $systems;
  }


  /**
    * getBooksChaptersVerses - Returns the books, chapters, verses for the given versification system.
    */
  public function getBooksChaptersVerses ($name)
  {
    $id = $this->getID ($name);
    $query = "SELECT book, chapter, verse FROM versification_data WHERE system = $id ORDER BY book, chapter, verse ASC;";
    $database_instance = Database_Instance::getInstance();
    $result = $database_instance->runQuery ($query);
    return $result;
  }
  


}



?>

File: gtk/src/versifications.cpp
/*
** Copyright (©) 2003-2013 Teus Benschop.
**  
** This program is free software; you can redistribute it and/or modify
** it under the terms of the GNU General Public License as published by
** the Free Software Foundation; either version 3 of the License, or
** (at your option) any later version.
**  
** This program is distributed in the hope that it will be useful,
** but WITHOUT ANY WARRANTY; without even the implied warranty of
** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
** GNU General Public License for more details.
**  
** You should have received a copy of the GNU General Public License
** along with this program; if not, write to the Free Software
** Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
**  
*/

#include "versifications.h"
#include "utilities.h"
#include "directories.h"
#include "gwrappers.h"
#include <libxml/xmlreader.h>
#include "bookdata.h"
#include "books.h"
#include "tiny_utilities.h"

Versification::Versification(const ustring & system_in, const ustring & filename)
{
  system = system_in;

  if (g_file_test(filename.c_str(), G_FILE_TEST_IS_REGULAR)) {
    gchar *contents;
    g_file_get_contents(filename.c_str(), &contents, NULL, NULL);
    xmlParserInputBufferPtr inputbuffer;
    inputbuffer = xmlParserInputBufferCreateMem(contents, strlen(contents), XML_CHAR_ENCODING_NONE);
    xmlTextReaderPtr reader = xmlNewTextReader(inputbuffer, NULL);
    if (reader) {
      char *opening_element = NULL;
      unsigned int myid;
      unsigned int mychapter;
      unsigned int myverse;

      while ((xmlTextReaderRead(reader) == 1)) {
        switch (xmlTextReaderNodeType(reader)) {
        case XML_READER_TYPE_ELEMENT:
          {
            opening_element = (char *)xmlTextReaderName(reader);
            if (!strcmp(opening_element, "triad")) {
              myid = 0;
              mychapter = 0;
              myverse = 0;
            }
            break;
          }
        case XML_READER_TYPE_TEXT:
          {
            char *text = (char *)xmlTextReaderValue(reader);
            if (text) {
              if (!strcmp(opening_element, "book")) {
                myid = books_english_to_id(text);
              }
              if (!strcmp(opening_element, "chapter")) {
                mychapter = convert_to_int(text);
              }
              if (!strcmp(opening_element, "verse")) {
                myverse = convert_to_int(text);
              }
              free(text);
            }
            break;
          }
        case XML_READER_TYPE_END_ELEMENT:
          {
            char *closing_element = (char *)xmlTextReaderName(reader);
            if (!strcmp(closing_element, "triad")) {
              if (myid == 0)
                break;
              if (mychapter == 0)
                break;
              if (myverse == 0)
                break;
              id.push_back(myid);
              chapter.push_back(mychapter);
              verse.push_back(myverse);
            }
            break;
          }
        }
      }
    }
    if (reader)
      xmlFreeTextReader(reader);
    if (inputbuffer)
      xmlFreeParserInputBuffer(inputbuffer);
    if (contents)
      g_free(contents);
  }
}

Versifications::Versifications(int dummy)
{
}

vector < ustring > Versifications::systems_get()
// Gets all available versification systems and store them in the object.
{
  if (available_systems.empty()) {
    // Get the system from the templates that come with Bibledit.
    ReadFiles rf1(Directories->get_package_data(), "versification", ".xml");
    for (unsigned int i = 0; i < rf1.files.size(); i++) {
      available_filenames.push_back(gw_build_filename(Directories->get_package_data(), rf1.files[i]));
      available_systems.push_back(filename_get_system(rf1.files[i]));
    }
    // Get the system from the templates provided by the user.
    ReadFiles rf2(Directories->get_templates_user(), "versification", ".xml");
    for (unsigned int i = 0; i < rf2.files.size(); i++) {
      available_filenames.push_back(gw_build_filename(Directories->get_templates_user(), rf2.files[i]));
      available_systems.push_back(filename_get_system(rf2.files[i]));
    }
    // Sort everything on name.
    quick_sort(available_systems, available_filenames, 0, available_systems.size());
  }
  return available_systems;
}

ustring Versifications::filename_get_system(ustring filename)
// Taking a filename, it produces the versification name from it.
{
  filename.erase(0, 14);
  filename.erase(filename.length() - 4, 4);
  replace_text(filename, "_", " ");
  size_t pos = 0;
  while (pos != string::npos) {
    filename.replace(pos, 1, upperCase(filename.substr(pos, 1)));
    pos = filename.find(" ", pos);
    if (pos != string::npos)
      pos++;
  }
  return filename;
}

unsigned int Versifications::system_pointer_get(ustring system)
/*
Get a pointer to the Versification for "system".
If "system" does not exist, it takes "English".
If "system" is not yet loaded, it first loads it, then gives the pointer.
*/
{
  // Load the localizations.
  if (available_systems.empty()) {
    systems_get();
  }
  // Is the requested system available?
  bool available = false;
  for (unsigned int i = 0; i < available_systems.size(); i++) {
    if (system == available_systems[i]) {
      available = true;
      break;
    }
  }
  // If the system is not available, take "English".
  if (!available)
    system = "English";
  // Get a pointer to the requested system, if it is already loaded.
  for (unsigned int i = 0; i < loaded_localizations.size(); i++) {
    if (system == loaded_localizations[i].system)
      return i;
  }
  // Load the requested system from the filename, and return a pointer to it.
  ustring filename;
  for (unsigned int i = 0; i < available_systems.size(); i++) {
    if (system == available_systems[i]) {
      filename = available_filenames[i];
      break;
    }
  }
  Versification versification(system, filename);
  loaded_localizations.push_back(versification);
  return loaded_localizations.size() - 1;
}

void Versifications::book2chaptersverses(const ustring & system, unsigned int id, vector < unsigned int >&chapters, vector < unsigned int >&verses)
{
  chapters.clear();
  verses.clear();
  unsigned int pointer = system_pointer_get(system);
  for (unsigned int i = 0; i < loaded_localizations[pointer].id.size(); i++) {
    if (id == loaded_localizations[pointer].id[i]) {
      chapters.push_back(loaded_localizations[pointer].chapter[i]);
      verses.push_back(loaded_localizations[pointer].verse[i]);
    }
  }
}
